core/src/test/java/org/apache/iceberg/TestMicroBatchBuilder.java

```diff
@@ -0,0 +1,163 @@
+  @Parameterized.Parameters
+  public static Object[][] parameters() {
+    return new Object[][] {
+        new Object[] { 1 },
+        new Object[] { 2 },
+    };
+  }
+
+  public TestMicroBatchBuilder(int formatVersion) {
+    super(formatVersion);
+  }
+
+  @Before
+  public void setupTableProperties() {
+    table.updateProperties().set(TableProperties.MANIFEST_MIN_MERGE_COUNT, "3").commit();
+  }
+
+  @Test
+  public void testGenerateMicroBatch() {
+    add(table.newAppend(), files("A", "B", "C", "D", "E"));
+
+    Snapshots.MicroBatch batch = Snapshots.from(table.snapshot(1L), table.io())
+        .specsById(table.specs())
+        .generate(0, Long.MAX_VALUE, true);
+    Assert.assertEquals(batch.snapshotId(), 1L);
+    Assert.assertEquals(batch.startFileIndex(), 0);
+    Assert.assertEquals(batch.endFileIndex(), 6);
+    Assert.assertEquals(batch.sizeInBytes(), 50);
+    Assert.assertTrue(batch.lastIndexOfSnapshot());
+    filesMatch(Lists.newArrayList("A", "B", "C", "D", "E"), filesToScan(batch.tasks()));
+
+    Snapshots.MicroBatch batch1 = Snapshots.from(table.snapshot(1L), table.io())
+        .specsById(table.specs())
+        .generate(0, 15L, true);
+    Assert.assertEquals(batch1.endFileIndex(), 1);
+    Assert.assertEquals(batch1.sizeInBytes(), 10);
+    filesMatch(Lists.newArrayList("A"), filesToScan(batch1.tasks()));
+
+    Snapshots.MicroBatch batch2 = Snapshots.from(table.snapshot(1L), table.io())
+        .specsById(table.specs())
+        .generate(batch1.endFileIndex(), 30L, true);
+    Assert.assertEquals(batch2.endFileIndex(), 4);
+    Assert.assertEquals(batch2.sizeInBytes(), 30);
+    filesMatch(Lists.newArrayList("B", "C", "D"), filesToScan(batch2.tasks()));
+
+    Snapshots.MicroBatch batch3 = Snapshots.from(table.snapshot(1L), table.io())
+        .specsById(table.specs())
+        .generate(batch2.endFileIndex(), 50L, true);
+    Assert.assertEquals(batch3.endFileIndex(), 6);
+    Assert.assertEquals(batch3.sizeInBytes(), 10);
+    filesMatch(Lists.newArrayList("E"), filesToScan(batch3.tasks()));
+  }
+
+  @Test
+  public void testGenerateMicroBatchWithSmallTargetSize() {
+    add(table.newAppend(), files("A", "B", "C", "D", "E"));
+
+    Snapshots.MicroBatch batch = Snapshots.from(table.snapshot(1L), table.io())
+        .specsById(table.specs())
+        .generate(0, 10L, true);
+    Assert.assertEquals(batch.snapshotId(), 1L);
+    Assert.assertEquals(batch.startFileIndex(), 0);
+    Assert.assertEquals(batch.endFileIndex(), 1);
+    Assert.assertEquals(batch.sizeInBytes(), 10);
+    Assert.assertFalse(batch.lastIndexOfSnapshot());
+    filesMatch(Lists.newArrayList("A"), filesToScan(batch.tasks()));
+
+    Snapshots.MicroBatch batch1 = Snapshots.from(table.snapshot(1L), table.io())
+        .specsById(table.specs())
+        .generate(batch.endFileIndex(), 5L, true);
+    Assert.assertEquals(batch1.endFileIndex(), 1);
+    Assert.assertEquals(batch1.sizeInBytes(), 0);
```

Comment:
Isn't it a problem to create batches with no files? I think that every batch where a file is available should contain at least one file to process, even if it is above the desired size. Otherwise, the job could get stuck and make no progress because a single file is larger than the batch size.

Reply:
Yes, that could lead to this problem, so typically I would suggest to set a relative large value to avoid this problem. My original thought is that we should strictly keep the batch size to not exceed the limitation. I can change to what you suggested.

Reply:
Changed.